From 9b556388e37efcc7d97953aae3549c319aaee319 Mon Sep 17 00:00:00 2001
From: Tianling Shen <cnsztl@gmail.com>
Date: Sun, 10 Aug 2025 13:38:52 +0800
Subject: [PATCH] mtd: spinand: add FM25S01A to fmsh driver

Add support for FudanMicro FM25S01A SPI NAND.
Datasheet: http://eng.fmsh.com/nvm/FM25S01A_ds_eng.pdf

According to datasheet, there is no QE (Quad Enable) bit for FM25S01A.
---
 drivers/mtd/nand/spi/fmsh.c | 49 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 49 insertions(+)

--- a/drivers/mtd/nand/spi/fmsh.c
+++ b/drivers/mtd/nand/spi/fmsh.c
@@ -40,6 +40,55 @@ static SPINAND_OP_VARIANTS(update_cache_variants,
 		SPINAND_PROG_LOAD_1S_1S_4S_OP(false, 0, NULL, 0),
 		SPINAND_PROG_LOAD_1S_1S_1S_OP(false, 0, NULL, 0));
 
+static int fm25s01a_ooblayout_ecc(struct mtd_info *mtd, int section,
+				 struct mtd_oob_region *region)
+{
+	return -ERANGE;
+}
+
+static int fm25s01a_ooblayout_free(struct mtd_info *mtd, int section,
+				  struct mtd_oob_region *region)
+{
+	if (section)
+		return -ERANGE;
+
+	region->offset = 2;
+	region->length = 62;
+
+	return 0;
+}
+
+static const struct mtd_ooblayout_ops fm25s01a_ooblayout = {
+	.ecc = fm25s01a_ooblayout_ecc,
+	.free = fm25s01a_ooblayout_free,
+};
+
 static const struct spinand_info fmsh_spinand_table[] = {
+	SPINAND_INFO("FM25S01A",
+		 SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0xE4),
+		 NAND_MEMORG(1, 2048, 64, 64, 1024, 20, 1, 1, 1),
+		 NAND_ECCREQ(1, 512),
+		 SPINAND_INFO_OP_VARIANTS(&read_cache_variants,
+				 &write_cache_variants,
+				 &update_cache_variants),
+		 0,
+		 SPINAND_ECCINFO(&fm25s01a_ooblayout, NULL)),
 };
 
 static const struct spinand_manufacturer_ops fmsh_spinand_manuf_ops = {
-- 
2.50.1
